File: go/160.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	curA, curB, cnt := headA, headB, 0
	for cnt < 3 {
		if curA == nil {
			curA = headB
			cnt++
		}
		if curB == nil {
			curB = headA
			cnt++
		}
		if curA == curB {
			return curA
		}
		curA = curA.Next
		curB = curB.Next
	}
	return nil
}

File: cpp/160.cpp

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
  ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
    if (!headA || !headB)
      return nullptr;
    ListNode *n{headA}, *m{headB};
    int cnt{0};
    while (cnt < 3) {
      if (!n)
        n = headB, ++cnt;
      if (!m)
        m = headA, ++cnt;
      if (n == m)
        return n;
      n = n->next, m = m->next;
    }
    return nullptr;
  }
};
